sdb/commands/stacks.py

```diff
@@ -216,120 +272,104 @@ def find_module_memory_segment(mod_name: str) -> Tuple[int, int]:
-                self.args.tstate = self.args.tstate.upper()
+            self.match_state = Stacks.resolve_state(self.args.tstate)
+            task_states = Stacks.TASK_STATES.values()
+            if self.match_state not in task_states:
+                valid_states = ", ".join(task_states)
+                raise sdb.CommandError(
+                    self.name, f"'{self.args.tstate}' is not a valid task state"
+                    f" (acceptable states: {valid_states})")
 
-        mod_start, mod_end = -1, -1
         if self.args.module:
-            mod_start, mod_size = Stacks.find_module_memory_segment(
+            if Stacks.find_module_memory_segment(self.args.module)[0] == -1:
+                raise sdb.CommandError(
+                    self.name,
+                    f"module '{self.args.module}' doesn't exist or isn't currently loaded"
+                )
+            self.mod_start, mod_size = Stacks.find_module_memory_segment(
                 self.args.module)
-            assert mod_start != -1
-            mod_end = mod_start + mod_size
-
+            assert self.mod_start != -1
+            self.mod_end = self.mod_start + mod_size
+
+    def match_stack(self, task: drgn.Object) -> bool:
+        task_state = Stacks.task_struct_get_state(task)
+        if self.args.tstate and self.match_state != task_state:
+            return False
+
+        frame_pcs = Stacks.get_frame_pcs(task)
+        mod_match = False
+        func_match = False
+        for frame_pc in frame_pcs:
+            if self.args.module and self.mod_start <= frame_pc < self.mod_end:
+                mod_match = True
+
+            if self.args.function and self.func_start <= frame_pc < self.func_end:
+                func_match = True
+
+        if (self.args.function and not func_match) or (self.args.module and
+                                                       not mod_match):
+            return False
+        return True
+
+    def print_header(self):
         header = "{:<18} {:<16s}".format("TASK_STRUCT", "STATE")
         if not self.args.all:
             header += " {:>6s}".format("COUNT")
         print(header)
         print("=" * 42)
 
-        #
-        # We inspect and group the tasks by recording their state and
-        # stack frames once in the following loop. We do this because
-        # on live systems state can change under us, thus running
-        # something like sdb.get_prog().stack_trace(task) twice (once for
-        # grouping and once for printing) could yield different stack
-        # traces resulting into misleading output.
-        #
+    #
+    # De-duplicate the objs (task_structs) using a dictionary indexed by
+    # task state and program counters. Return a collection sorted by number
+    # of tasks per stack.
+    #
+    @staticmethod
+    def aggregate_stacks(
+            objs: Iterable[drgn.Object]) -> List[Tuple[Any, List[drgn.Object]]]:
         stack_aggr: Dict[Any, List[drgn.Object]] = defaultdict(list)
```

Comment:
[nit] since stack_key is now a fixed-length tuple, each entry of known type (yay!), can we be more specific than `Any` on these 2 lines?  Would `Tuple[str, Tuple[int]]` work?  (not sure how the variable-length Tuple-of-ints should be specified, it may have to fall back on Any)

Reply:
I looked this up, looks like we can do something similar:

>  Tuples can also be used as immutable, varying-length sequences. You can use the type Tuple[T, ...] (with a literal ... – it’s part of the syntax) for this purpose

So I can do `Tuple[str, Tuple[int, ...]]`